### PR TITLE
Adds support for kwargs and default arguments in the python client, and improves how parameter information is displayed in the "view API" page

### PR DESCRIPTION
**Pull request by abidlabs**
This PR adds support for kwargs and default arguments in the Client. In other words, suppose you have the following demo:

```py
def calculator(num1, operation, num2):
    if operation == "add":
        return num1 + num2
    elif operation == "subtract":
        return num1 - num2
    elif operation == "multiply":
        return num1 * num2
    elif operation == "divide":
        if num2 == 0:
            raise gr.Error("Cannot divide by zero!")
        return num1 / num2

demo = gr.Interface(
    calculator,
    ["number", gr.Radio(["add", "subtract", "multiply", "divide"], value="add), "number"],
    "number",
    examples=[
        [5, "add", 3],
        [4, "divide", 2],
        [-4, "multiply", 2.5],
        [0, "subtract", 1.2],
    ],
).launch()
```

When you connect to this app via the client, you can now make queries like this:

```py
from gradio_client import Client

client = Client("http://127.0.0.1:7861/")
result = client.predict(
		num1=3,
		operation="add",
		num2=10,
		api_name="/predict"
)
print(result)
```

Note that parameters to `.predict()` are now supplied with key-word arguments. For backwards compatibility, the old approach using positional arguments continues to work. e.g. 

```py
result = client.predict(
		3,
		"add",
		33,
		api_name="/predict"
)
```

but the key-word approach is encouraged because it is more robust and also allows us to take advantage of default arguments. In the example above, we could also do:

```py
from gradio_client import Client

client = Client("http://127.0.0.1:7861/")
result = client.predict(
		num1=33,
		num2=33,
		api_name="/predict"
)
print(result)
```

taking advantage of the fact that `operation` is `"add"` by default, as specified in the initial value of the component.

Closes: https://github.com/gradio-app/gradio/issues/3540
Closes: https://github.com/gradio-app/gradio/issues/5344
Closes: https://github.com/gradio-app/gradio/issues/4498 

Finally, I've updated the design of the view API page to showcase all this information about each parameter. (parameter names, whether it is required, or the default fallback value if it is not). Here's a partial screenshot:

image

### CODE REVIEW

File: gradio/blocks.py
Comment (freddyaboulton) [inline gradio/blocks.py:2518]: The default value may not be json-serializable. We should check against that return `None` in that case.

**Comment by freddyaboulton** [comment]
> I would say that this also closes: https://github.com/gradio-app/gradio/issues/4498 because it would be confusing to show two different "defaults" -- the function default and the slider default. Instead, we just show the function default if it is provided.

I think it would be better if we showed the component default value. That's the standard way gradio app developers specify default values for the app. I don't think function defaults are that common in gradio since they don't really take effect. We also know the component value is serializable since it's been postprocessed by the component and is in the config.

**Comment by abidlabs** [comment]
> I think it would be better if we showed the component default value. That's the standard way gradio app developers specify default values for the app. I don't think function defaults are that common in gradio since they don't really take effect. We also know the component value is json-serializable since it's been postprocessed by the component and is in the config.

Interesting. So just to to understand, you're suggesting using the initial value for a component, if it has one, as the default when accessing via the clients?

We could do that, but what if a developer wants to set a default for the clients but not change their UI?

**Comment by freddyaboulton** [comment]
Yes that's what I was suggesting! The other point about using the python function's default value is that I'm not sure it would work for some file based components (like Image). If you have a python function that is expecting a PIL image, you'd have to set the default value of the parameter to be a URL string for it to be useful in the ViewAPI page, which is awkward. 

I hear you about maybe wanting to specify a different default value for the API as opposed to the UI but I think pulling the defaults from the component as opposed to the function is closer to what an app author would expect (and saves them from having to update defaults in two places if they do want to use the UI default for the API default).

**Comment by abidlabs** [comment]
Yeah makes sense, I'll change that!

Also just noticed this issue where the author suggests exactly your approach so that's good validation too: https://github.com/gradio-app/gradio/issues/5344

**Comment by abidlabs** [comment]
By the way @gary149 if you have feedback on the new design of the view API page, that would be super helpful.

Here's an example of how it looks: https://gradio-pr-deploys-pr-7732-all-demos.hf.space/demo/altair_plot/?__theme=dark&view=api

**Comment by freddyaboulton** [comment]
@abidlabs Yea good point about default value of None. Agree with your proposal!

**Comment by abidlabs** [comment]
Ok I've made the change @freddyaboulton. Still working on the documentation, but should otherwise be good for a review

File: test/test_utils.py
Comment (freddyaboulton) [inline test/test_utils.py:521]: Why doesn't the b get picked up in this case?

File: js/app/src/api_docs/CodeSnippet.svelte
Comment (freddyaboulton) [inline js/app/src/api_docs/CodeSnippet.svelte:63]: nit: there is an extra space added before the comma

Pasted image 20240321162808

Comment (abidlabs) [reply js/app/src/api_docs/CodeSnippet.svelte:63]: It looks like that but I think there's no space actually, maybe its a padding issue

File: test/test_utils.py
Comment (abidlabs) [reply test/test_utils.py:521]: We stop as soon as we hit an *args or **kwargs because we no longer know for sure which component that parameter corresponds to

File: guides/08_gradio-clients-and-lite/01_getting-started-with-the-python-client.md
Comment (freddyaboulton) [inline guides/08_gradio-clients-and-lite/01_getting-started-with-the-python-client.md:161]: Nit: there's a double ` in after None

File: client/python/gradio_client/client.py
Comment (freddyaboulton) [inline client/python/gradio_client/client.py:681]: I think we need to traverse the value here to only pull out the actual filepath or url

File: js/app/src/api_docs/CodeSnippet.svelte
Comment (freddyaboulton) [reply js/app/src/api_docs/CodeSnippet.svelte:63]: Interesting. I think ideal would be no padding but not a big deal

**Comment by abidlabs** [comment]
Ok the docs should be ready as well now!

File: client/python/gradio_client/client.py
Comment (abidlabs) [reply client/python/gradio_client/client.py:681]: So the view API page takes care of this automatically:

image

but the `.view_api()` function does not

image

```py
import gradio as gr

demo = gr.Interface(
    lambda x:x,
    gr.Audio("test.mp3"),
    "audio"
)

_, url, _ = demo.launch()

from gradio_client import Client

client = Client(url)

client.view_api()
```

Comment (abidlabs) [reply client/python/gradio_client/client.py:681]: I'll fix in the `.view_api()`

**Review by freddyaboulton — COMMENTED** [review]
Very slick PR @abidlabs ! Tested a variety of demos and they all work well!

one thing that would be nice is to have the URL of the file if it's on spaces otherwise the file upload would fail?

image

File: client/python/gradio_client/client.py
Comment (freddyaboulton) [reply client/python/gradio_client/client.py:681]: Yea sorry that's what I meant!

**Comment by abidlabs** [comment]
Thanks @freddyaboulton! Agree will figure out a better solution for files

**Comment by abidlabs** [comment]
Ok that should address everything in the review:

image

I'll merge in later today unless @aliabd has any other suggestions

**Review by aliabd — APPROVED** [review]
Really like this @abidlabs! 

I have two small suggestions that you can make the final call on. 

* I feel like the multiple parameters are blending together visually. I know you have the faint orange separator (in light mode), maybe making it more prominent or increasing the space? 
* I think 'Defaults to _' and 'Required' should occupy the same position. Right now 'Required' is after the name and type, where as the default value is in the end of the description. I think it would be better if you maybe put them both at the end of the first line, on the right.

Everything else looks so clean!

**Comment by abidlabs** [comment]
Thanks for the reviews @freddyaboulton and @aliabd. I agree with your suggestions @aliabd, will make the changes and merge

**Comment by abidlabs** [comment]
Updated. It now looks like this (more margin between the parameters and defaults on the same line as the required): 

image

**Pull request merged**
